programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -130,55 +132,69 @@ impl<'a, 'b> CallerAccount<'a, 'b> {
         // account_info points to host memory. The addresses used internally are
         // in vm space so they need to be translated.
         let lamports = {
-            // Double translate lamports out of RefCell
-            let ptr = translate_type::<u64>(
+            // Triple translate lamports out of Rc<RefCell>
+            let ptr_box = translate_type::<VmBoxOfRefCell<u64>>(
                 memory_mapping,
-                account_info.lamports.as_ptr() as u64,
+                account_info.lamports.addr,
                 invoke_context.get_check_aligned(),
             )?;
             if direct_mapping {
                 check_account_info_pointer(
                     invoke_context,
-                    *ptr,
+                    ptr_box.value,
                     account_metadata.vm_lamports_addr,
                     "lamports",
                 )?;
             }
-            translate_type_mut::<u64>(memory_mapping, *ptr, invoke_context.get_check_aligned())?
+
+            translate_type_mut::<u64>(
+                memory_mapping,
+                ptr_box.value,
+                invoke_context.get_check_aligned()
+            )?
         };
 
         let owner = translate_type_mut::<Pubkey>(
             memory_mapping,
-            account_info.owner as *const _ as u64,
+            account_info.owner,
             invoke_context.get_check_aligned(),
         )?;
 
         let (serialized_data, vm_data_addr, ref_to_len_in_vm) = {
             // Double translate data out of RefCell
-            let data = *translate_type::<&[u8]>(
+            let ptr_box = translate_type_mut::<VmBoxOfRefCell<VmSlice<u8>>>(
                 memory_mapping,
-                account_info.data.as_ptr() as *const _ as u64,
+                account_info.data.addr,
                 invoke_context.get_check_aligned(),
             )?;
+
+            let vm_data_addr = ptr_box.value.ptr(); // virtual address of VmSlice's contents
+
             if direct_mapping {
                 check_account_info_pointer(
                     invoke_context,
-                    data.as_ptr() as u64,
+                    vm_data_addr,
                     account_metadata.vm_data_addr,
                     "data",
                 )?;
             }
 
+            // Translate the vmSlice into a physically addressed "true" Rust slice
+            let data_slice = ptr_box.value.translate_mut(
+                memory_mapping,
+                invoke_context.get_check_aligned()
+            )?;
+
             consume_compute_meter(
                 invoke_context,
-                (data.len() as u64)
+                (data_slice.len() as u64)
                     .checked_div(invoke_context.get_compute_budget().cpi_bytes_per_unit)
                     .unwrap_or(u64::MAX),
             )?;
 
+            let len_offset = size_of::<u64>().saturating_mul(4) as u64;
```

Comment:
Original code uses implicit pointer magic to get the virtual address of the slice, and then adds 8 so that it holds the vaddr of the slice's length field (which is worthy of raw C code, not Rust!). I can't really do that without defining a bunch of as_ptr() sorts of things specifically for this purpose, so I just replaced their hack with a version of the same hack without the hidden pointer arithmetic. Should probably add the comment that they didn't, but basically it's still just the vaddr of the VmSlice's length field.

Reply:
Thanks, the comment in code is helpful context!